diff --git a/init/main.c b/init/main.c
index 4aa7b04b48f1..b979094e836a 100644
--- a/init/main.c
+++ b/init/main.c
@@ -563,6 +563,8 @@ asmlinkage __visible void __init start_kernel(void)
 	}
 
 
+	/* parameters may set static keys */
+	jump_label_init();
 	parse_early_param();
 	after_dashes = parse_args("Booting kernel",
 				  static_command_line, __start___param,
@@ -572,8 +574,6 @@ asmlinkage __visible void __init start_kernel(void)
 		parse_args("Setting init args", after_dashes, NULL, 0, -1, -1,
 			   NULL, set_init_arg);
 
-	jump_label_init();
-
 	/*
 	 * These use large bootmem allocations and must precede
 	 * kmem_cache_init()
