diff --git a/fs/binfmt_elf.c b/fs/binfmt_elf.c
index 3cc79d2c430b..7911009074df 100644
--- a/fs/binfmt_elf.c
+++ b/fs/binfmt_elf.c
@@ -827,6 +827,7 @@ static int load_elf_binary(struct linux_binprm *bprm)
 		current->flags |= PF_RANDOMIZE;
 
 	setup_new_exec(bprm);
+	install_exec_creds(bprm);
 
 	/* Do this so that we can load the interpreter, if need be.  We will
 	   change some of these later */
@@ -1028,7 +1029,6 @@ static int load_elf_binary(struct linux_binprm *bprm)
 
 	kfree(elf_phdata);
 
-	install_exec_creds(bprm);
 	retval = create_elf_tables(bprm, &loc->elf_ex,
 			  load_addr, interp_load_addr);
 	if (retval < 0)
