diff --git a/fs/f2fs/data.c b/fs/f2fs/data.c
index 7b5f0e54a1e1..f758ef520bef 100644
--- a/fs/f2fs/data.c
+++ b/fs/f2fs/data.c
@@ -568,6 +568,9 @@ static struct bio *f2fs_grab_read_bio(struct inode *inode, block_t blkaddr,
 	struct bio_post_read_ctx *ctx;
 	unsigned int post_read_steps = 0;
 
+	if (!f2fs_is_valid_blkaddr(sbi, blkaddr, DATA_GENERIC))
+		return ERR_PTR(-EFAULT);
+
 	bio = f2fs_bio_alloc(sbi, min_t(int, nr_pages, BIO_MAX_PAGES), false);
 	if (!bio)
 		return ERR_PTR(-ENOMEM);
