diff --git a/sound/usb/mixer.c b/sound/usb/mixer.c
index e87ff963e3af..c6a8b3328a39 100644
--- a/sound/usb/mixer.c
+++ b/sound/usb/mixer.c
@@ -1060,6 +1060,14 @@ static void volume_control_quirks(struct usb_mixer_elem_info *cval,
 						(cval->control << 8) | 2, 7936);
 		break;
 
+	case USB_ID(0x0495, 0x3042): /* ESS Technology Asus USB DAC */
+		if ((strstr(kctl->id.name, "Playback Volume") != NULL) ||
+			strstr(kctl->id.name, "Capture Volume") != NULL) {
+			cval->min >>= 8;
+			cval->max = 0;
+			cval->res = 1;
+		}
+		break;
 	}
 }
 
