diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index f31ab36181c4..63f7561332e2 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -6961,6 +6961,7 @@ static inline int find_idlest_cpu(struct sched_domain *sd, struct task_struct *p
 
 #ifdef CONFIG_SCHED_SMT
 DEFINE_STATIC_KEY_FALSE(sched_smt_present);
+EXPORT_SYMBOL_GPL(sched_smt_present);
 
 static inline void set_idle_cores(int cpu, int val)
 {
