diff --git a/mm/memory.c b/mm/memory.c
index d5e7e83fefdd..86b250b72fca 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -1828,7 +1828,7 @@ static int insert_pfn(struct vm_area_struct *vma, unsigned long addr,
 				goto out_unlock;
 			}
 			entry = pte_mkyoung(*pte);
-			entry = maybe_mkwrite(pte_mkdirty(entry), vma);
+			entry = maybe_mkwrite(pte_mkdirty(entry), vma->vm_flags);
 			if (ptep_set_access_flags(vma, addr, pte, entry, 1))
 				update_mmu_cache(vma, addr, pte);
 		}
