diff --git a/drivers/misc/mnh/mnh-pcie.c b/drivers/misc/mnh/mnh-pcie.c
index 02597e66542c..4c0c09147246 100644
--- a/drivers/misc/mnh/mnh-pcie.c
+++ b/drivers/misc/mnh/mnh-pcie.c
@@ -937,7 +937,8 @@ int mnh_sg_build(void *dmadest, size_t size, struct mnh_sg_entry **sg,
 
 	down_read(&current->mm->mmap_sem);
 	n_num = get_user_pages(current, current->mm,
-			(unsigned long) dmadest, p_num, 1, 1, sgl->mypage,
+			(unsigned long) dmadest, p_num,
+			FOLL_WRITE | FOLL_FORCE, sgl->mypage,
 				NULL);
 	up_read(&current->mm->mmap_sem);
 	if (n_num < 0) {
