diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 897698005ea2..ec55ab0c7aff 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -4332,9 +4332,8 @@ static int __sched_setscheduler(struct task_struct *p,
 	int queue_flags = DEQUEUE_SAVE | DEQUEUE_MOVE;
 	struct rq *rq;
 
-	/* may grab non-irq protected spin_locks */
-	if (pi)
-		BUG_ON(in_interrupt());
+	/* The pi code expects interrupts enabled */
+	BUG_ON(pi && in_interrupt());
 recheck:
 	/* double check policy once rq lock held */
 	if (policy < 0) {
