diff --git a/drivers/clk/qcom/clk-rcg2.c b/drivers/clk/qcom/clk-rcg2.c
index c38859fde3ac..fb55c5e93c5a 100644
--- a/drivers/clk/qcom/clk-rcg2.c
+++ b/drivers/clk/qcom/clk-rcg2.c
@@ -149,7 +149,7 @@ static int update_config(struct clk_rcg2 *rcg, u32 cfg)
 	WARN_CLK(hw->core, name, count == 0,
 			"rcg didn't update its configuration.");
 
-	return 0;
+	return -EBUSY;
 }
 
 static int clk_rcg2_set_parent(struct clk_hw *hw, u8 index)
