diff --git a/kernel/time/alarmtimer.c b/kernel/time/alarmtimer.c
index a5daa945ed2a..03cb21520d33 100644
--- a/kernel/time/alarmtimer.c
+++ b/kernel/time/alarmtimer.c
@@ -102,9 +102,9 @@ static int alarmtimer_rtc_add_device(struct device *dev,
 				struct class_interface *class_intf)
 {
 	unsigned long flags;
-	int err = 0;
 	struct rtc_device *rtc = to_rtc_device(dev);
 	struct wakeup_source *__ws;
+	int ret = 0;
 
 	if (rtcdev)
 		return -EBUSY;
@@ -117,13 +117,13 @@ static int alarmtimer_rtc_add_device(struct device *dev,
 	spin_lock_irqsave(&rtcdev_lock, flags);
 	if (!rtcdev) {
 		if (!try_module_get(rtc->owner)) {
-			spin_unlock_irqrestore(&rtcdev_lock, flags);
-			return -1;
+			ret = -1;
+			goto unlock;
 		}
 
-		err = rtc_irq_register(rtc, &alarmtimer_rtc_task);
-		if (err)
-			goto rtc_irq_reg_err;
+		ret = rtc_irq_register(rtc, &alarmtimer_rtc_task);
+		if (ret)
+			goto unlock;
 
 		rtcdev = rtc;
 		/* hold a reference so it doesn't go away */
@@ -131,12 +131,12 @@ static int alarmtimer_rtc_add_device(struct device *dev,
 		ws = __ws;
 		__ws = NULL;
 	}
-
-rtc_irq_reg_err:
+unlock:
 	spin_unlock_irqrestore(&rtcdev_lock, flags);
 
 	wakeup_source_unregister(__ws);
-	return err;
+
+	return ret;
 }
 
 static void alarmtimer_rtc_remove_device(struct device *dev,
