diff --git a/drivers/net/wireless/ath/ath10k/hw.c b/drivers/net/wireless/ath/ath10k/hw.c
index 558214cef688..5b7faab1cf02 100644
--- a/drivers/net/wireless/ath/ath10k/hw.c
+++ b/drivers/net/wireless/ath/ath10k/hw.c
@@ -488,7 +488,7 @@ const struct ath10k_hw_values qca6174_values = {
 
 const struct ath10k_hw_values qca99x0_values = {
 	.pdev_suspend_option		= WMI_PDEV_SUSPEND_AND_DISABLE_INTR,
-	.rtc_state_val_on		= 5,
+	.rtc_state_val_on		= 7,
 	.ce_count			= 12,
 	.msi_assign_ce_max		= 12,
 	.num_target_ce_config_wlan	= 10,
