diff --git a/drivers/of/Kconfig b/drivers/of/Kconfig
index 60efdd584d15..ea02d4af4b7b 100644
--- a/drivers/of/Kconfig
+++ b/drivers/of/Kconfig
@@ -118,6 +118,10 @@ config OF_OVERLAY
 	  While this option is selected automatically when needed, you can
 	  enable it manually to improve device tree unit test coverage.
 
+config OF_DMA_DEFAULT_COHERENT
+	# arches should select this if DMA is coherent by default for OF devices
+	bool
+
 config OF_BATTERYDATA
 	def_bool y
 	help
