diff --git a/drivers/usb/dwc3/core.h b/drivers/usb/dwc3/core.h
index 8d1a8d9d3132..d2620daeb9d8 100644
--- a/drivers/usb/dwc3/core.h
+++ b/drivers/usb/dwc3/core.h
@@ -970,6 +970,7 @@ struct dwc3_scratchpad_array {
  * 	1	- -3.5dB de-emphasis
  * 	2	- No de-emphasis
  * 	3	- Reserved
+ * @dis_metastability_quirk: set to disable metastability quirk.
  * @err_evt_seen: previous event in queue was erratic error
  * @usb3_u1u2_disable: if true, disable U1U2 low power modes in Superspeed mode
  * @in_lpm: indicates if controller is in low power mode (no clocks)
@@ -1167,6 +1168,8 @@ struct dwc3 {
 	bool			b_suspend;
 	unsigned int		vbus_draw;
 
+	unsigned		dis_metastability_quirk:1;
+
 	u16			imod_interval;
 	struct workqueue_struct *dwc_wq;
 	struct work_struct      bh_work;
