diff --git a/drivers/staging/android/ion/ion_priv.h b/drivers/staging/android/ion/ion_priv.h
index 117e51581fc6..039467b2de3f 100644
--- a/drivers/staging/android/ion/ion_priv.h
+++ b/drivers/staging/android/ion/ion_priv.h
@@ -619,6 +619,9 @@ int ion_walk_heaps(struct ion_client *client, int heap_id,
 		   unsigned int type, void *data,
 		   int (*f)(struct ion_heap *heap, void *data));
 
+struct ion_handle *ion_handle_get_by_id_nolock(struct ion_client *client,
+					       int id);
+
 int ion_handle_put(struct ion_handle *handle);
 
 void show_ion_usage(struct ion_device *dev);
