diff --git a/include/net/tcp.h b/include/net/tcp.h
index ca0bc35e33c8..15fe430bc893 100644
--- a/include/net/tcp.h
+++ b/include/net/tcp.h
@@ -380,7 +380,7 @@ extern int tcp_proc_delayed_ack_control(struct ctl_table *table, int write,
 				void __user *buffer, size_t *length,
 				loff_t *ppos);
 
-void tcp_enter_quickack_mode(struct sock *sk);
+void tcp_enter_quickack_mode(struct sock *sk, unsigned int max_quickacks);
 static inline void tcp_dec_quickack_mode(struct sock *sk,
 					 const unsigned int pkts)
 {
